signimage/signimage.go

```diff
@@ -0,0 +1,352 @@
+package main
+
+import (
+	//"context"
+	"archive/tar"
+	"bytes"
+	"errors"
+	"flag"
+	"fmt"
+	"github.com/docker/cli/cli/config"
+	"io"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	//"github.com/docker/cli/cli/config/configfile"
+	dockertypes "github.com/docker/cli/cli/config/types"
+	"github.com/google/go-containerregistry/pkg/authn"
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/go-containerregistry/pkg/v1/mutate"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/google/go-containerregistry/pkg/v1/remote/transport"
+	"github.com/google/go-containerregistry/pkg/v1/tarball"
+)
+
+func getenv(key string, fallback string) string {
+	value, exists := os.LookupEnv(key)
+	if !exists {
+		value = fallback
+	}
+	return value
+}
+
+func checkarg(arg *string, envvar string, fallback string) {
+	if *arg == "" {
+		if envvar != "" {
+			v, exists := os.LookupEnv(envvar)
+			if exists {
+				*arg = v
+			} else if !exists && fallback != "" {
+				*arg = fallback
+			} else {
+				fmt.Printf("%s not found:\n", envvar)
+				flag.PrintDefaults()
+				os.Exit(0)
+			}
+		}
+	}
+	return
+}
+
+func canonicalisePath(path string) string {
+	canonical := strings.Replace("/"+path, "/./", "/", -1)
```

Comment:
since the files in the layer are in relative path, isn't it enough just to add a "/" to the start of the path of the file?

Reply:
The files in a layer are normally relative but I think their exact format depends how exactly they were created. They are sometimes stored as  'path/` and sometimes `./path/` and sometimes aren't relative at all but are absolute (I suspect this is in baselayers) which become `//path` with a "/" appended. 

In the CRD we are asking for absolute paths, so this function is just doing some search and replace to clean up those patterns which then allows us to do string comparison on the paths to easily check if the file we have in the image is one to sign